### DISCUSSION TITLE
How to properly remotely shut down Unium

### DISCUSSION THREAD

**Discussion by bryandickson** [gwaredd/unium#106] [Q&A]
I'm working on a "restartApp" function that shuts down all gameobjects and reloads the main scene.  This causes Unium to throw a bunch of http/socket errors for in-flight requests.  How could I properly shut down the server (from a web hook or remotely called component function), before restarting the app?

**Comment by gwaredd** [comment]
There are start and stop server functions on the Unium component, which will be called automatically when the component is enabled and disabled.

Stopping the server closes the listening socket and any in-flight requests should respond with a 500 Internal Server error. Unium is multithreaded so the response should work in parallel to the listener shutting down. If the thread pool is being restarted as part of the restart then it's possible the worker threads are abnormally killing the client socket before the data is sent and the connection closed. It's also possible there's a bug as closing down the server hasn't been extensively tested :) It would require some investigation.

**Reply by bryandickson** [reply]
I think that is what's happening.  Here's some info that might help: 

This code: 

```
public static void Restart(string startScene = "Main")
{
	GameObject go = new GameObject("DestroyableRoot");
	Object.DontDestroyOnLoad(go);

	foreach (var root in go.scene.GetRootGameObjects())
	{
		if (!dontDestroy.Contains(root.name))
		{
			Debug.Log("Destroying " + root.name);
			Object.DestroyImmediate(root);
		}
	}

	// wait a few secs
	System.Threading.Thread.Sleep(3000);
	
	Debug.Log("Loading scene: " + startScene);
	SceneManager.LoadScene(startScene, LoadSceneMode.Single);
}
```

gives the error: 

`HttpResponseException: Exception of type 'gw.proto.http.HttpResponseException' was thrown.
gw.proto.http.HttpResponse.WriteHeaders () (at Assets/{redacted}/Plugins/Unium/Core/gw.proto.http/HttpResponse.cs:106)
gw.proto.http.HttpResponse.Reject (gw.proto.http.ResponseCode code) (at Assets/{redacted}/Plugins/Unium/Core/gw.proto.http/HttpResponse.cs:149)
gw.proto.http.HttpRequest.Reject (gw.proto.http.ResponseCode code) (at Assets/{redacted}/Plugins/Unium/Core/gw.proto.http/HttpRequest.cs:79)
gw.unium.RequestAdapterHTTP.Reject (gw.proto.http.ResponseCode code) (at Assets/{redacted}/Plugins/Unium/Routing/RouterAdapters.cs:50)
gw.unium.Route.Dispatch (gw.unium.RequestAdapter request) (at Assets/{redacted}/Plugins/Unium/Routing/Router.cs:41)
UniumComponent.ProcessWebRequestsOnGameThread () (at Assets/{redacted}/Plugins/Unium/UniumComponent.cs:246)
UniumComponent.LateUpdate () (at Assets/{redacted}/Plugins/Unium/UniumComponent.cs:178)
`

**Reply by gwaredd** [reply]
The component is trying to update the request in `LateUpdate()` which curious ...

Firstly - has the Unium component been destroyed at this point? If so, why is LateUpdate being called? was StopServer() called when it was destroyed? and if it was why isn't mQueuedRequests empty? I guess there is a theoretical race condition if a new connection is being processed before the server is stopped but I would expect the code to be robust enough to handle that and the connection just timeout.

Secondly ... there are two exceptions here. The first exception is being through in the routing/handler code, being caught at [`Router.cs:41`](https://github.com/gwaredd/unium/blob/e08aef88d56d01fa4764da6b27dd7d108e6b4094/Assets/Unium/Routing/Router.cs#L41), which should elicit a 400 Bad Request. The [second exception is thrown](https://github.com/gwaredd/unium/blob/e08aef88d56d01fa4764da6b27dd7d108e6b4094/Assets/Unium/Core/gw.proto.http/HttpResponse.cs#L106) during that because it thinks that either the connection is closed or that it has already sent a response.

Quick fix would be to put a try/catch around `Router.cs:41`. 

Better fix would be to look at why the first exception is being thrown and why LateUpdate() is being called.
